fix(bootstrap): Add one time reset of user bootstrap node list

anthonybilinski opened the pull request:
- [x] Commits follow our [git commit guidelines](https://github.com/qTox/qTox/blob/master/CONTRIBUTING.md#git-commit-guidelines)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qtox/qtox/6465)
<!-- Reviewable:end -->

sudden6 commented:
do you think it makes sense to version the settings too? I have not seen this elsewhere, but I can't come up with a good reason not to do it.

anthonybilinski committed:
feat(Settings): Add system for versioning and upgrading Settings

Similar to how History handles SCHEMA_VERSION. Run separately on global and
personal settings, since some state in global, and personal settings can’t be
done globally since they require the passkey.

Restrict a user from downgrading past the saved settings version, due to
possible compatibility breaks or old qTox versions re-introducing corrupt state
that was already healed.

Pass in new profile state for personal settings rather than relying on settings
file presence because personal settings can be stored in either the personal
settings file or global settings file. This was introduced in
aea9eea8a4e2f95a7175954183959c1070bcc315 when personal settings were first
moved to their own file.

anthonybilinski committed:
fix(core): One time reset of user bootstrap node list to default

Due to poor design, bootstrapNodes.json was saved the same if manually set by
user, or if just storing qTox defaults. Because of this, for users using
defaults, they would be stuck with stale defaults indefinitely without this one
time reset.

The user list that's removed is backed up to bootstrapNodes.backup.json.

Going forward, the user list and default list will be stored separately.

Fix #6452

anthonybilinski committed:
feat(core): Store default bootstrap list separate from user list

Allows qTox to tell the difference between a default list that should be
upgraded when defaults are changed, ora user list that should never be changed.

anthonybilinski reviewed (commented):
On reconsideration, yes I think that's best and I've changed it to work similar to `History`.

Upgrading is handled by Settings now too, rather than exposing the bug mitigation state to different classes.

> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6465#-:-MxGKLiaEwQTPJZs27C_:b-g37uar)* status: 0 of 1 LGTMs obtained



<!-- Sent from Reviewable.io -->

sphaerophoria commented on src/persistence/settings.cpp:
```diff
@@ -112,16 +117,34 @@ void Settings::loadGlobal()
     QSettings s(filePath, QSettings::IniFormat);
     s.setIniCodec("UTF-8");
 
+    s.beginGroup("Version");
```
Why do you have a group with a single value? Is this just how ini's have to be formatted?

sphaerophoria commented on src/persistence/paths.cpp:
```diff
@@ -365,11 +365,25 @@ QString Paths::getAppCacheDirPath() const
+QString Paths::getExampleNodesFilePath() const
+{
+    QDir dir(getSettingsDirPath());
+    constexpr static char nodesFileName[] = "bootstrapNodes.example.json";
```
This is a nice solution. Gives someone a reference without preventing us from updating it later.

sphaerophoria commented on src/persistence/globalsettingsupgrader.cpp:
```diff
@@ -0,0 +1,79 @@
+        const auto& paths = settings.getPaths();
+
+        QFile badFile{paths.getUserNodesFilePath()};
+        if (badFile.exists()) {
```
Can we prompt the user for this? No one is going to see the qWarning

anthonybilinski reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6465)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/settings.cpp, line 132 at r4](https://reviewable.io/reviews/qtox/qtox/6465#-MwP2HaK32e0anhnqnvN-r4-132:-MxHwF3A3xeenxck1AtO:b5xk9u4) ([raw file](https://github.com/qtox/qtox/blob/4febe25c27eeec2f9e5b49f0ca6bf70e97b77726/src/persistence/settings.cpp#L132)):*
<details><summary><i>Previously, sphaerophoria wrote…</i></summary><blockquote>

Why do you have a group with a single value? Is this just how ini's have to be formatted?
</blockquote></details>

It's semi-required by the file format, see bottom of this section: https://doc.qt.io/qt-5/qsettings.html#Format-enum

If it was not explicitly put in a group in code, it would be under "General" in the file, which doesn't seem much different.
___
*[src/persistence/globalsettingsupgrader.cpp, line 33 at r6](https://reviewable.io/reviews/qtox/qtox/6465#-MxGwIuv2LeMZ955T49Y-r6-33:-MxHwa-p9ciWcLznQZJH:b-sjmqje) ([raw file](https://github.com/qtox/qtox/blob/867c21b7fabdaa19dd0bc33e85d9fb4c177d9d12/src/persistence/globalsettingsupgrader.cpp#L33)):*
<details><summary><i>Previously, sphaerophoria wrote…</i></summary><blockquote>

Can we prompt the user for this? No one is going to see the qWarning
</blockquote></details>

We don't have a GUI at time of global settings load thanks to a mess of initialization order which wouldn't be easy to fix =/

I agree no one's going to see this initially, but they might look into logs or open an issue if qTox refuses to start, which this causes. Same issue in `Settings` where `GlobalSettingsUpgrader::doUpgrade` is called.


<!-- Sent from Reviewable.io -->

sudden6 reviewed (changes requested):
> Reviewed 50 of 59 files at r4, 8 of 9 files at r7, 4 of 4 files at r8, 4 of 4 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6465)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/personalsettingsupgrader.cpp, line 26 at r7](https://reviewable.io/reviews/qtox/qtox/6465#-MxTuV0FBKdBaMY90J67:-MxTuV0FBKdBaMY90J68:btb4xej) ([raw file](https://github.com/qtox/qtox/blob/77d7d4c96b2bf1011af6643565b2fce2f9ac76fd/src/persistence/personalsettingsupgrader.cpp#L26)):*
> ```cpp
> #include <cassert>
> 
> bool PersonalSettingsUpgrader::doUpgrade(SettingsSerializer& settingsSerializer, int fromVer, int toVer)
> ```

same, make these `unsigned`
___
*[src/persistence/settings.h, line 720 at r7](https://reviewable.io/reviews/qtox/qtox/6465#-MxTuZp34Ji8qdiwGvXt:-MxTuZp34Ji8qdiwGvXu:bb7od0r) ([raw file](https://github.com/qtox/qtox/blob/77d7d4c96b2bf1011af6643565b2fce2f9ac76fd/src/persistence/settings.h#L720)):*
> ```objc
>     static QThread* settingsThread;
>     Paths paths;
>     int globalSettingsVersion = 0;
> ```

same
___
*[src/persistence/settings.cpp, line 493 at r7](https://reviewable.io/reviews/qtox/qtox/6465#-MxTupq7-uIiDhiDWpUh:-MxTupq7-uIiDhiDWpUi:b-r8h8u3) ([raw file](https://github.com/qtox/qtox/blob/77d7d4c96b2bf1011af6643565b2fce2f9ac76fd/src/persistence/settings.cpp#L493)):*
> ```cpp
>     bool defaultSettings = false;
>     if (!QFile(filePath).exists()) {
>         defaultSettings = true;
> ```

move to the `else` path below
___
*[src/persistence/globalsettingsupgrader.cpp, line 26 at r7](https://reviewable.io/reviews/qtox/qtox/6465#-MxTuG8g4xn1wm9NeKy0:-MxTuG8g4xn1wm9NeKy1:b-3hznmd) ([raw file](https://github.com/qtox/qtox/blob/77d7d4c96b2bf1011af6643565b2fce2f9ac76fd/src/persistence/globalsettingsupgrader.cpp#L26)):*
> ```cpp
> #include <cassert>
> 
> bool GlobalSettingsUpgrader::doUpgrade(Settings& settings, int fromVer, int toVer)
> ```

if you make these `unsigned int` you can probably remove the casts and I don't expect there will be negative versions.


<!-- Sent from Reviewable.io -->

sudden6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6465)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/personalsettingsupgrader.cpp, line 26 at r7](https://reviewable.io/reviews/qtox/qtox/6465#-MxTuV0FBKdBaMY90J67:-MxUCciABGxTJHAVHc5x:b7caqk4) ([raw file](https://github.com/qtox/qtox/blob/77d7d4c96b2bf1011af6643565b2fce2f9ac76fd/src/persistence/personalsettingsupgrader.cpp#L26)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

same, make these `unsigned`
</blockquote></details>

Retracted, based on out-of-band discussions with regards to Qt size types compatibility.


<!-- Sent from Reviewable.io -->

sudden6 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6465)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/settings.cpp, line 493 at r7](https://reviewable.io/reviews/qtox/qtox/6465#-MxTupq7-uIiDhiDWpUh:-MxUGzSX9pQJp_3Mh4U0:b-60qyt5) ([raw file](https://github.com/qtox/qtox/blob/77d7d4c96b2bf1011af6643565b2fce2f9ac76fd/src/persistence/settings.cpp#L493)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

move to the `else` path below
</blockquote></details>

Sorry, I misread that and merging is not possible, but it can be converted to a oneliner: `const  bool defaultSettings = !QFile(filePath).exists()`


<!-- Sent from Reviewable.io -->

anthonybilinski reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6465)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/settings.cpp, line 493 at r7](https://reviewable.io/reviews/qtox/qtox/6465#-MxTupq7-uIiDhiDWpUh:-MxUCeJJ128pCGa9Iw1J:b-1xgh2e) ([raw file](https://github.com/qtox/qtox/blob/77d7d4c96b2bf1011af6643565b2fce2f9ac76fd/src/persistence/settings.cpp#L493)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

Sorry, I misread that and merging is not possible, but it can be converted to a oneliner: `const  bool defaultSettings = !QFile(filePath).exists()`
</blockquote></details>

arg I was thinking to avoid the case where they have a global settings file but no personal settings file, but this would miss every new profile for a single user. Running it only if it's a new profile settings file could potentially miss the upgrade if a user has a global settings file but no user settings file though.. which looks like a possible case based on this loading behaviour, and digging through history I found that that was in aea9eea8a4e2f95a7175954183959c1070bcc315. So it looks like we have no perfect way to tell if it's a new profile based on Settings files.

That should really be done through an upgrade itself at some point...

For now though, I'll just pass in if it's a new profile from Profile to Settings::loadPersonal, since Profile is aware and is always doing the loading.


<!-- Sent from Reviewable.io -->

sudden6 reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 8 of 8 files at r10, 4 of 4 files at r11, 4 of 4 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6465#-:-MxUdsb92tIB_0zF47vf:b-uwkowd)* status: :shipit: complete! 1 of 1 LGTMs obtained



<!-- Sent from Reviewable.io -->
